### PR TITLE
libbpf-tools: add CO-RE readahead

### PR DESCRIPTION
**Pull request by ethercflow**
The output looks like:

```sh
Tracing fs read-ahead ... Hit Ctrl-C to end.
^C
Readahead unsued/total pages: 96/473288
     msecs               : count    distribution
         0 -> 1          : 473191   |****************************************|
         2 -> 3          : 1        |                                        |
```

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

### CODE REVIEW

**Comment by ethercflow** [comment]
Hi @anakryiko, another tool PTAL when you have time, thank you. :)

File: libbpf-tools/readahead.bpf.c [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct page *);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} birth SEC(".maps");
+
+static struct hist hist;
+
+SEC("fentry/__do_page_cache_readahead")
+int BPF_PROG(fentry____do_page_cache_readahead)
```
Comment (anakryiko) [inline libbpf-tools/readahead.bpf.c:30]: btw, you don't have to folow this convention of duplicating program type in function name, it's actually even better to have shorter function name, so maybe try just `do_page_cache_readahead`. bpftool will show that it's an fentry program anyways.

File: libbpf-tools/readahead.bpf.c [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+{
+	u32 pid = bpf_get_current_pid_tgid();
+
+	bpf_map_update_elem(&in_readahead, &pid, &pid, 0);
```
Comment (anakryiko) [inline libbpf-tools/readahead.bpf.c:34]: value type is `u8`, yet you are trying to put pid there?

File: libbpf-tools/readahead.bpf.c [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	return 0;
+}
+
+SEC("fexit/__page_cache_alloc")
+int BPF_PROG(fexit____page_cahe_alloc, gfp_t gfp, struct page *ret)
```
Comment (anakryiko) [inline libbpf-tools/readahead.bpf.c:39]: typo: cache

File: libbpf-tools/readahead.bpf.c [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+{
+	u32 pid = bpf_get_current_pid_tgid(), ts;
```
Comment (anakryiko) [inline libbpf-tools/readahead.bpf.c:41]: ts has to be u64

File: libbpf-tools/readahead.c [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	obj = readahead_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF ojbect\n");
+		return 1;
+	}
+
+	err = readahead_bpf__load(obj);
+	if (err) {
```
Comment (anakryiko) [inline libbpf-tools/readahead.c:109]: you can combine into a single open_and_load() step to keep it nice and short

File: libbpf-tools/readahead.c [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	signal(SIGINT, sig_handler);
+
+	printf("Tracing fs read-ahead ... Hit Ctrl-C to end.\n");
+
+	sleep(env.duration);
+	printf("\n");
+
+	histp = &obj->bss->hist;
+
+	printf("Readahead unsued/total pages: %d/%d\n",
```
Comment (anakryiko) [inline libbpf-tools/readahead.c:129]: typo: unused

File: libbpf-tools/readahead.bpf.c (hunk 4 above)
Comment (ethercflow) [reply libbpf-tools/readahead.bpf.c:41]: Oh so sorry, I changed `u64 id` to `u32 pid`, but forgot to check ts's type.

File: libbpf-tools/readahead.bpf.c (hunk 2 above)
Comment (ethercflow) [reply libbpf-tools/readahead.bpf.c:34]: Yes, this is my intention. I only care about the existence of the key in the map, but not the val value. So use u8 to save memory.

File: libbpf-tools/readahead.c (hunk 6 above)
Comment (ethercflow) [reply libbpf-tools/readahead.c]: Fixed, thanks!

File: libbpf-tools/readahead.c (hunk 5 above)
Comment (ethercflow) [reply libbpf-tools/readahead.c]: Fixed, thanks!

File: libbpf-tools/readahead.bpf.c (hunk 4 above)
Comment (ethercflow) [reply libbpf-tools/readahead.bpf.c]: Fixed, thanks!

File: libbpf-tools/readahead.bpf.c (hunk 3 above)
Comment (ethercflow) [reply libbpf-tools/readahead.bpf.c]: Fixed, thanks!

File: libbpf-tools/readahead.bpf.c (hunk 1 above)
Comment (ethercflow) [reply libbpf-tools/readahead.bpf.c]: Got it, thanks!

**Comment by ethercflow** [comment]
Hi @anakryiko! Thanks for the review, pushed fixes for all your comments. :)

File: libbpf-tools/readahead.bpf.c (hunk 2 above)
Comment (anakryiko) [reply libbpf-tools/readahead.bpf.c:34]: hm... so the value doesn't matter, ok. But it's not clear, so I wouldn't do it, I'd rather do `u64 one = 1;` and use that as a value. Also, value size internally in kernel is rounded up to 8 bytes anyways, so you are not saving anything, really.

Comment (ethercflow) [reply libbpf-tools/readahead.bpf.c:34]: Got it, thanks! I'll change it to make it clear.

**Comment by ethercflow** [comment]
Hi @yonghong-song , PTAL when you have time, thanks! :)

**Comment by ethercflow** [comment]
Hi @yonghong-song , PTAL when you have time, thanks! :)

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Pull request merged**